# Input and Output
- The `read_csv` function supports dozens of parameters for handling various file formats.
- All datasets are in the `io_examples` folder.

In [1]:
import pandas as pd

- Some datasets may not have a header column. Pandas will treat the first row as the colum names.
- Passing `header=None` creates a numeric index for the columns.
- The `names` parameter can set custom column names.

In [2]:
pd.read_csv("io_examples/sensor_data.csv")
pd.read_csv("io_examples/sensor_data.csv", header=None)
pd.read_csv(
    "io_examples/sensor_data.csv",
    header=None,
    names=["timestamp", "temperature", "humidity"],
)

,timestamp,temperature,humidity
0,2024-01-15 06:00,22.1,45.3
1,2024-01-15 07:00,21.8,46.1
2,2024-01-15 08:00,23.4,44.0
3,2024-01-15 09:00,25.0,42.7
4,2024-01-15 10:00,26.3,41.2
5,2024-01-15 11:00,27.8,39.8
6,2024-01-15 12:00,28.5,38.5


- Datasets may have comments or blank lines before the actual data.
- The `skiprows` parameter skips a specified number of lines at the top of the file.

In [3]:
pd.read_csv("io_examples/quarterly_report.csv", skiprows=3)

,Region,Q1,Q2,Q3,Q4
0,North,52000,61000,58000,73000
1,South,48000,53000,55000,62000
2,East,61000,59000,64000,71000
3,West,45000,50000,52000,68000


- By default, pandas recognizes common missing-value markers like `NaN`, `NA`, and empty strings.
- Use `na_values` to define additional values that qualify as missing values.

In [4]:
pd.read_csv("io_examples/customer_data.csv", na_values=["--", "missing"])

,name,age,zipcode,satisfaction,referral_source
0,Alice Johnson,34.0,2134,9.0,Google
1,Bob Smith,NaN,10001,7.0,Friend
2,Carol White,28.0,90210,NaN,Instagram
3,David Lee,45.0,501,8.0,NaN
4,Emma Davis,NaN,60601,10.0,Google
5,Frank Miller,52.0,30301,NaN,Twitter
6,Grace Kim,29.0,2134,6.0,Friend
7,Henry Chen,41.0,10002,NaN,NaN


- The `dtype` parameter sets column types on import.
- The dataset loses the leading zeroes for the `zipcode` column (like `00501`). Conversion later can't help.
- Pass the `dtype` parameter a dictionary with column names as keys and data types as values.

In [5]:
pd.read_csv(
    "io_examples/customer_data.csv", na_values=["--", "missing"], dtype={"zipcode": str}
)

,name,age,zipcode,satisfaction,referral_source
0,Alice Johnson,34.0,02134,9.0,Google
1,Bob Smith,NaN,10001,7.0,Friend
2,Carol White,28.0,90210,NaN,Instagram
3,David Lee,45.0,00501,8.0,NaN
4,Emma Davis,NaN,60601,10.0,Google
5,Frank Miller,52.0,30301,NaN,Twitter
6,Grace Kim,29.0,02134,6.0,Friend
7,Henry Chen,41.0,10002,NaN,NaN


- The `nrows` parameter imports only the first `n` rows.
- The parameter can preview a large dataset without loading the complete filev into memory.
- The `index_col` parameter sets a column's values as the index labels.

In [6]:
pd.read_csv("fortune1000.csv", nrows=5)
pd.read_csv("fortune1000.csv", nrows=5, index_col="Rank")

,Company,Sector,Industry,Revenue,Profits,Employees
Rank,,,,,,
1,Walmart,Retailing,General Merchandisers,482130,14694,2300000
2,Exxon Mobil,Energy,Petroleum Refining,246204,16150,75600
3,Apple,Technology,"Computers, Office Equipment",233715,53394,110000
4,Berkshire Hathaway,Financials,Insurance: Property and Casualty (Stock),210821,24083,331000
5,McKesson,Health Care,Wholesalers: Health Care,181241,1476,70400


## Export DataFrame to CSV File
- The `to_csv` method exports a `DataFrame` to a CSV file.
- Its first argument is the filename.
- By default, pandas will include the index. Set the `index` parameter to False to exclude the index.
- The `columns` parameter limits the exported columns.

In [7]:
url = "https://data.cityofnewyork.us/api/views/25th-nujf/rows.csv"

In [8]:
baby_names = pd.read_csv(url)
baby_names.head()

,Year of Birth,Gender,Ethnicity,Child's First Name,Count,Rank
0,2021,FEMALE,ASIAN AND PACIFIC ISLANDER,Chloe,71,1
1,2021,FEMALE,ASIAN AND PACIFIC ISLANDER,Olivia,71,1
2,2021,FEMALE,ASIAN AND PACIFIC ISLANDER,Emma,66,2
3,2021,FEMALE,ASIAN AND PACIFIC ISLANDER,Mia,59,3
4,2021,FEMALE,ASIAN AND PACIFIC ISLANDER,Ava,53,4


In [9]:
for column in ["Gender", "Ethnicity", "Child's First Name"]:
    baby_names[column] = baby_names[column].str.strip().str.title()

In [10]:
baby_names.head()

,Year of Birth,Gender,Ethnicity,Child's First Name,Count,Rank
0,2021,Female,Asian And Pacific Islander,Chloe,71,1
1,2021,Female,Asian And Pacific Islander,Olivia,71,1
2,2021,Female,Asian And Pacific Islander,Emma,66,2
3,2021,Female,Asian And Pacific Islander,Mia,59,3
4,2021,Female,Asian And Pacific Islander,Ava,53,4


In [11]:
baby_names.to_csv(
    "baby_names.csv",
    index=False,
    columns=["Year of Birth", "Child's First Name", "Count"],
)

## Import Excel Workbook into pandas
- The `read_excel` function reads an Excel file/workbook into a `DataFrame`.

In [12]:
pd.read_excel("Data - Single Worksheet.xlsx")

,First Name,Last Name,City,Gender
0,Brandon,James,Miami,M
1,Sean,Hawkins,Denver,M
2,Judy,Day,Los Angeles,F
3,Ashley,Ruiz,San Francisco,F
4,Stephanie,Gomez,Portland,F


- Use the `sheet_name` parameter if the workbook contains multiple worksheets.
- Pass a single worksheet name or a list of worksheet names/index positions.
- Pandas will store multiple worksheets in a Python dictionary.
- The keys will be the worksheet names, and the values will be the `DataFrame` objects.

In [13]:
pd.read_excel("Data - Multiple Worksheets.xlsx")
pd.read_excel("Data - Multiple Worksheets.xlsx", sheet_name="Data 1")
pd.read_excel("Data - Multiple Worksheets.xlsx", sheet_name="Data 2")
pd.read_excel("Data - Multiple Worksheets.xlsx", sheet_name=0)
pd.read_excel("Data - Multiple Worksheets.xlsx", sheet_name=1)

,First Name,Last Name,City,Gender
0,Parker,Power,Raleigh,F
1,Preston,Prescott,Philadelphia,F
2,Ronaldo,Donaldo,Bangor,M
3,Megan,Stiller,San Francisco,M
4,Bustin,Jieber,Austin,F


In [14]:
pd.read_excel("Data - Multiple Worksheets.xlsx", sheet_name=["Data 1", "Data 2"])
type(pd.read_excel("Data - Multiple Worksheets.xlsx", sheet_name=["Data 1", "Data 2"]))
pd.read_excel("Data - Multiple Worksheets.xlsx", sheet_name=[0, 1])

{0:   First Name Last Name           City Gender
 0    Brandon     James          Miami      M
 1       Sean   Hawkins         Denver      M
 2       Judy       Day    Los Angeles      F
 3     Ashley      Ruiz  San Francisco      F
 4  Stephanie     Gomez       Portland      F,
 1:   First Name Last Name           City Gender
 0     Parker     Power        Raleigh      F
 1    Preston  Prescott   Philadelphia      F
 2    Ronaldo   Donaldo         Bangor      M
 3      Megan   Stiller  San Francisco      M
 4     Bustin    Jieber         Austin      F}

- Pass the `sheet_name` parameter an argument of `None` to include all worksheets.
- The `keys` method on a dictionary returns a collection of its keys.
- The `values` method on a dictionary returns a collection of its values.

In [15]:
data = pd.read_excel("Data - Multiple Worksheets.xlsx", sheet_name=None)
data

{'Data 1':   First Name Last Name           City Gender
 0    Brandon     James          Miami      M
 1       Sean   Hawkins         Denver      M
 2       Judy       Day    Los Angeles      F
 3     Ashley      Ruiz  San Francisco      F
 4  Stephanie     Gomez       Portland      F,
 'Data 2':   First Name Last Name           City Gender
 0     Parker     Power        Raleigh      F
 1    Preston  Prescott   Philadelphia      F
 2    Ronaldo   Donaldo         Bangor      M
 3      Megan   Stiller  San Francisco      M
 4     Bustin    Jieber         Austin      F}

In [16]:
data.keys()

dict_keys(['Data 1', 'Data 2'])

In [17]:
data["Data 1"]
data["Data 2"]

,First Name,Last Name,City,Gender
0,Parker,Power,Raleigh,F
1,Preston,Prescott,Philadelphia,F
2,Ronaldo,Donaldo,Bangor,M
3,Megan,Stiller,San Francisco,M
4,Bustin,Jieber,Austin,F


In [18]:
pd.concat(data.values())
pd.concat(data.values(), ignore_index=True)

,First Name,Last Name,City,Gender
0,Brandon,James,Miami,M
1,Sean,Hawkins,Denver,M
2,Judy,Day,Los Angeles,F
3,Ashley,Ruiz,San Francisco,F
4,Stephanie,Gomez,Portland,F
5,Parker,Power,Raleigh,F
6,Preston,Prescott,Philadelphia,F
7,Ronaldo,Donaldo,Bangor,M
8,Megan,Stiller,San Francisco,M
9,Bustin,Jieber,Austin,F


## Export Excel Workbook from pandas
- The `read_excel` function writes a `DataFrame` to an Excel workbook.
- By default, the worksheet name will be `Sheet1`.
- The `sheet_name` parameter customizes the worksheet name.
- By default, pandas will keep the index as a column. Set `index=False` to exclude the index.
- The `columns` parameter limits the columns that pandas will write.

In [19]:
url = "https://data.cityofnewyork.us/api/views/25th-nujf/rows.csv"
baby_names = pd.read_csv(url)
baby_names.head()

,Year of Birth,Gender,Ethnicity,Child's First Name,Count,Rank
0,2021,FEMALE,ASIAN AND PACIFIC ISLANDER,Chloe,71,1
1,2021,FEMALE,ASIAN AND PACIFIC ISLANDER,Olivia,71,1
2,2021,FEMALE,ASIAN AND PACIFIC ISLANDER,Emma,66,2
3,2021,FEMALE,ASIAN AND PACIFIC ISLANDER,Mia,59,3
4,2021,FEMALE,ASIAN AND PACIFIC ISLANDER,Ava,53,4


In [20]:
males = baby_names[baby_names["Gender"] == "MALE"]
females = baby_names[baby_names["Gender"] == "FEMALE"]

In [21]:
females.to_csv("NYC Baby Data.csv")
females.to_excel("NYC Baby Data.xlsx")
females.to_excel("NYC Baby Data.xlsx", index=False)
females.to_excel("NYC Baby Data.xlsx", index=False, sheet_name="Females")
females.to_excel(
    "NYC Baby Data.xlsx",
    index=False,
    sheet_name="Females",
    columns=["Year of Birth", "Child's First Name", "Rank"],
)

- The `ExcelWriter` class writes one or more `DataFrame` objects to an Excel file.
- Use a context manager (the `with` keyword) in combination with the `ExcelWriter` object and an assigned variable.
- Invoke the `to_excel` method on every `DataFrame` to include in the Excel workbook and pass in the `ExcelWriter` object as the first argument.
- The `to_excel` method supports `sheet_name`, `index`, and `columns` parameters.

In [22]:
with pd.ExcelWriter("NYC Baby Data.xlsx") as excel_file:
    females.to_excel(
        excel_file,
        sheet_name="Females",
        index=False,
        columns=["Year of Birth", "Child's First Name", "Rank"],
    )
    males.to_excel(excel_file, sheet_name="Males", index=False)